File: server/api/story_hooks.py
"""Author-managed chapter-ending hook ledger."""

from datetime import datetime
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy.ext.asyncio import AsyncSession

from api.auth import ProjectPermission, get_current_user, verify_project_permission
from db.models_core import User
from db.models_long_generation import StoryHook
from db.session import get_db
from services.story_hooks import (
    StoryHookConflictError,
    StoryHookNotFoundError,
    StoryHookValidationError,
    create_story_hook,
    list_story_hooks,
    update_story_hook,
)

router = APIRouter()
HookStatus = Literal["open", "deferred", "resolved", "abandoned"]


class StoryHookCreateRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    source_chapter_id: str = Field(alias="sourceChapterId", min_length=1, max_length=32)
    hook_type: str = Field(alias="hookType", min_length=1, max_length=50)
    concrete_event: str = Field(alias="concreteEvent", min_length=1, max_length=4000)
    unresolved_question: str = Field(alias="unresolvedQuestion", min_length=1, max_length=2000)
    payoff_by_chapter: int | None = Field(None, alias="payoffByChapter", ge=1)


class StoryHookUpdateRequest(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")

    expected_revision: int = Field(alias="expectedRevision", ge=1)
    hook_type: str | None = Field(None, alias="hookType", min_length=1, max_length=50)
    concrete_event: str | None = Field(None, alias="concreteEvent", min_length=1, max_length=4000)
    unresolved_question: str | None = Field(
        None,
        alias="unresolvedQuestion",
        min_length=1,
        max_length=2000,
    )
    payoff_by_chapter: int | None = Field(None, alias="payoffByChapter", ge=1)
    payoff_chapter_id: str | None = Field(None, alias="payoffChapterId", max_length=32)
    status: HookStatus | None = None
    resolution: str | None = Field(None, max_length=4000)


class StoryHookResponse(BaseModel):
    id: str
    project_id: str
    source_chapter_id: str
    payoff_chapter_id: str | None
    hook_type: str
    concrete_event: str
    unresolved_question: str
    payoff_by_chapter: int | None
    status: HookStatus
    novelty_signature: str
    resolution: str
    revision: int
    created_at: datetime
    updated_at: datetime


def _response(hook: StoryHook) -> StoryHookResponse:
    return StoryHookResponse.model_validate(hook, from_attributes=True)


def _raise_story_hook_error(error: Exception) -> None:
    if isinstance(error, StoryHookNotFoundError):
        raise HTTPException(status_code=404, detail={"code": "STORY_HOOK_NOT_FOUND"}) from error
    if isinstance(error, StoryHookConflictError):
        raise HTTPException(
            status_code=409,
            detail={
                "code": "STORY_HOOK_REVISION_CONFLICT",
                "currentRevision": error.current_revision,
            },
        ) from error
    if isinstance(error, StoryHookValidationError):
        raise HTTPException(
            status_code=422,
            detail={"code": "STORY_HOOK_INVALID", "message": str(error)},
        ) from error
    raise error


@router.get("/projects/{project_id}/hooks", response_model=list[StoryHookResponse])
async def get_story_hooks(
    project_id: str,
    status: list[HookStatus] | None = Query(default=None),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[StoryHookResponse]:
    await verify_project_permission(project_id, ProjectPermission.VIEW, user, db)
    hooks = await list_story_hooks(db, project_id=project_id, statuses=set(status or []))
    return [_response(hook) for hook in hooks]


@router.post("/projects/{project_id}/hooks", response_model=StoryHookResponse, status_code=201)
async def post_story_hook(
    project_id: str,
    request: StoryHookCreateRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> StoryHookResponse:
    await verify_project_permission(project_id, ProjectPermission.MANAGE_OUTLINE, user, db)
    try:
        hook = await create_story_hook(
            db,
            project_id=project_id,
            **request.model_dump(by_alias=False),
        )
        await db.commit()
        await db.refresh(hook)
    except Exception as error:
        await db.rollback()
        _raise_story_hook_error(error)
    return _response(hook)


@router.patch("/projects/{project_id}/hooks/{hook_id}", response_model=StoryHookResponse)
async def patch_story_hook(
    project_id: str,
    hook_id: str,
    request: StoryHookUpdateRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> StoryHookResponse:
    await verify_project_permission(project_id, ProjectPermission.MANAGE_OUTLINE, user, db)
    changes = request.model_dump(
        by_alias=False,
        exclude={"expected_revision"},
        exclude_unset=True,
    )
    try:
        hook = await update_story_hook(
            db,
            hook_id=hook_id,
            project_id=project_id,
            expected_revision=request.expected_revision,
            changes=changes,
        )
        await db.commit()
        await db.refresh(hook)
    except Exception as error:
        await db.rollback()
        _raise_story_hook_error(error)
    return _response(hook)

File: app/src/api/story-hooks.ts
import { delay, request, USE_MOCK } from './http'
import type { StoryHook, StoryHookCreate, StoryHookPatch, StoryHookStatus } from '@/types'

interface StoryHookDto {
  id: string
  project_id: string
  source_chapter_id: string
  payoff_chapter_id: string | null
  hook_type: string
  concrete_event: string
  unresolved_question: string
  payoff_by_chapter: number | null
  status: StoryHookStatus
  novelty_signature: string
  resolution: string
  revision: number
  created_at: string
  updated_at: string
}

function fromDto(row: StoryHookDto): StoryHook {
  return {
    id: row.id,
    projectId: row.project_id,
    sourceChapterId: row.source_chapter_id,
    payoffChapterId: row.payoff_chapter_id ?? undefined,
    hookType: row.hook_type,
    concreteEvent: row.concrete_event,
    unresolvedQuestion: row.unresolved_question,
    payoffByChapter: row.payoff_by_chapter ?? undefined,
    status: row.status,
    noveltySignature: row.novelty_signature,
    resolution: row.resolution,
    revision: row.revision,
    createdAt: row.created_at,
    updatedAt: row.updated_at
  }
}

const localHooks = new Map<string, StoryHook[]>()

function localList(projectId: string) {
  const hooks = localHooks.get(projectId) ?? []
  localHooks.set(projectId, hooks)
  return hooks
}

function cloneHooks(hooks: StoryHook[]) {
  return structuredClone(hooks)
}

export const storyHooksApi = {
  async list(projectId: string, statuses: StoryHookStatus[] = []): Promise<StoryHook[]> {
    if (USE_MOCK) {
      await delay(50)
      const hooks = localList(projectId)
      return cloneHooks(statuses.length ? hooks.filter((hook) => statuses.includes(hook.status)) : hooks)
    }
    const query = new URLSearchParams()
    statuses.forEach((status) => query.append('status', status))
    const suffix = query.size ? `?${query.toString()}` : ''
    return (await request<StoryHookDto[]>(`/projects/${projectId}/hooks${suffix}`)).map(fromDto)
  },

  async create(projectId: string, input: StoryHookCreate): Promise<StoryHook> {
    if (USE_MOCK) {
      await delay(70)
      const now = new Date().toISOString()
      const hook: StoryHook = {
        id: `hook-${Date.now().toString(36)}`,
        projectId,
        sourceChapterId: input.sourceChapterId,
        hookType: input.hookType,
        concreteEvent: input.concreteEvent,
        unresolvedQuestion: input.unresolvedQuestion,
        payoffByChapter: input.payoffByChapter,
        status: 'open',
        noveltySignature: '',
        resolution: '',
        revision: 1,
        createdAt: now,
        updatedAt: now
      }
      localList(projectId).push(hook)
      return structuredClone(hook)
    }
    return fromDto(await request<StoryHookDto>(`/projects/${projectId}/hooks`, {
      method: 'POST',
      body: JSON.stringify(input)
    }))
  },

  async update(projectId: string, hookId: string, patch: StoryHookPatch): Promise<StoryHook> {
    if (USE_MOCK) {
      await delay(70)
      const hook = localList(projectId).find((item) => item.id === hookId)
      if (!hook) throw new Error('story_hook_not_found')
      if (hook.revision !== patch.expectedRevision) throw new Error('story_hook_revision_conflict')
      const { expectedRevision: _expectedRevision, ...changes } = patch
      Object.assign(hook, changes, {
        payoffByChapter: changes.payoffByChapter === null ? undefined : changes.payoffByChapter ?? hook.payoffByChapter,
        payoffChapterId: changes.payoffChapterId === null ? undefined : changes.payoffChapterId ?? hook.payoffChapterId,
        revision: hook.revision + 1,
        updatedAt: new Date().toISOString()
      })
      return structuredClone(hook)
    }
    return fromDto(await request<StoryHookDto>(`/projects/${projectId}/hooks/${hookId}`, {
      method: 'PATCH',
      body: JSON.stringify(patch)
    }))
  }
}
